src/libstore/derived-path-map.hh

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include "types.hh"
+#include "derived-path.hh"
+
+namespace nix {
+
+/**
+ * A simple Trie, of sorts. Conceptually a map of `SingleDerivedPath` to
+ * values.
+ *
+ * Concretely, an n-ary tree, as described below. A
+ * `SingleDerivedPath::Opaque` maps to the value of an immediate child
+ * of the root node. A `SingleDerivedPath::Built` maps to a deeper child
+ * node: the `SingleDerivedPath::Built::drvPath` is first mapped to a a
+ * child node (inductively), and then the
+ * `SingleDerivedPath::Built::output` is used to look up that child's
+ * child via its map. In this manner, every `SingleDerivedPath` is
+ * mapped to a child node.
```

Comment:
```suggestion
 * mapped to a child node.
 *
 * @param V   A type to instantiate for each output. For example, the scheduler uses `DerivedPathMap<CreateDerivationAndRealiseGoal>` to remember which goals correspond to which outputs.
```

Something like this would be useful. Perhaps you can phrase it more accurately.